File: Mixins-scala/src/clothingFactory/ClothingFactory.scala
package clothingFactory

import common.Item
import events.Event

import scala.jdk.CollectionConverters._

abstract class ClothingFactory {
  val catalog: Array[String] = new Array[String](10)
  val clothings: Map[String, List[String]] = Map()
  var numberOfClothings: Integer = 0

  def makeClothing(clothingName: String): Item = {
    var item: Item = new Item(clothingName, "cheap", adjustItemList(clothings(clothingName)).toArray)
    Thread.sleep(getPrecisionCoefficient)
    wrap(item)
    item
  }

  def addClothingSelection(clothingName: String, standardItems: List[String]): Unit = {
    catalog(numberOfClothings) = clothingName
    clothings + (clothingName -> standardItems)
    numberOfClothings += 1
  }

  def adjustItemList(items: List[String]) : List[String]
  def getPrecisionCoefficient: Long
  def wrap(item: Item): Unit
  def getProductionType: String
}

File: Strategy&Template-java/src/main/java/templateMethod/ClothingFactory.java
package templateMethod;

import common.*;
import enums.Clothings;
import enums.Types;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class ClothingFactory {

    private Map<Clothings, List<Material>> clothings = new HashMap<Clothings, List<Material>>();
    private List<Clothings> catalog = new ArrayList<Clothings>();
    List<Event> teamBuildingEvents = new ArrayList<>();

    public Clothing makeClothing(Clothings name) throws InterruptedException {
        Clothing clothing = new Clothing(name, getType());
        clothing.setMaterials(adjustMaterialList(clothings.get(name)));
        Thread.sleep(getPrecisionCoefficient());
        wrap(clothing);
        return clothing;
    }

    public void addCatalogSelection(Clothings name, List<Material> standardMaterials) {
        clothings.put(name, standardMaterials);
        catalog.add(name);
    }

    public List<Clothings> getCatalog() {
        return catalog;
    }

    abstract Types getType();

    abstract List<Material> adjustMaterialList(List<Material> materials);

    abstract long getPrecisionCoefficient();

    abstract void wrap(Clothing clothing);

    public abstract void organizeEvent(int budget, int numberOfAttendees);

    public List<Event> getEvents() {
        return teamBuildingEvents;
    }
}
